src/interpreter/mod.rs

```diff
@@ -600,20 +604,28 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
                     Misc => {
                         let src = self.eval_operand(operand)?;
                         let src_ty = self.operand_ty(operand);
-                        // FIXME(solson): Wrong for almost everything.
-                        warn!("misc cast from {:?} to {:?}", src_ty, dest_ty);
-                        let dest_size = self.type_size(dest_ty);
-                        let src_size = self.type_size(src_ty);
-                        let dest_align = self.type_align(dest_ty);
-
-                        // Hack to support fat pointer -> thin pointer casts to keep tests for
-                        // other things passing for now.
-                        let is_fat_ptr_cast = pointee_type(src_ty).map_or(false, |ty| !self.type_is_sized(ty));
-
-                        if dest_size == src_size || is_fat_ptr_cast {
-                            self.memory.copy(src, dest, dest_size, dest_align)?;
+                        if self.type_is_immediate(src_ty) {
+                            // FIXME: dest_ty should already be monomorphized
+                            let dest_ty = self.monomorphize(dest_ty, self.substs());
+                            assert!(self.type_is_immediate(dest_ty));
+                            let src_val = self.read_primval(src, src_ty)?;
+                            let dest_val = self.cast_primval(src_val, dest_ty)?;
+                            self.memory.write_primval(dest, dest_val)?;
                         } else {
-                            return Err(EvalError::Unimplemented(format!("can't handle cast: {:?}", rvalue)));
+                            // Casts from a fat-ptr.
+                            assert!(self.type_is_fat_ptr(src_ty));
```

Comment:
`type_is_fat_ptr` should be the value to check for this `if`/`else`, not `type_is_immediate`.